File: deploy/cloudflare/src/pano.js
/** VISION-matching Street View view plan and ephemeral thumbnail decode. */

import jpegJs from "jpeg-js";
import { FACE_COUNT, FACE_SIZE, FACES_BYTES } from "./model.js";

export const THUMB_SIZE = 64;
export const FACE_FOV = 90;
export const THUMBNAIL_ENDPOINT = "https://geo0.ggpht.com/cbk";
export const VIEW_USER_AGENT = "VISION-Community-view/1";
export const QUERY_VIEW_CAP = 4;
export const STREET_LEASE_CAP = {
  scene: { slow: 1, medium: 4, max: 8 },
  object: { slow: 1, medium: 2, max: 4 },
};
export const CLI_LEASE_CAP = {
  scene: { slow: 16, medium: 64, max: 128 },
  object: { slow: 8, medium: 32, max: 64 },
};

const INVENTED_PREFIXES = ["synthetic:", "Prototype", "CommunityPano", "wikimedia:"];

export function usesStreetViews(panoId) {
  if (typeof panoId !== "string" || !panoId.trim()) return false;
  const identity = panoId.trim();
  if (identity.includes("maps.googleapis.com") || identity.startsWith("http")) return false;
  return !INVENTED_PREFIXES.some((prefix) => identity.startsWith(prefix));
}

export function wrapHeading(heading) {
  const value = Number(heading) || 0;
  return ((value % 360) + 360) % 360;
}

export function thumbnailFov(zoom) {
  const z = Number.isFinite(Number(zoom)) ? Number(zoom) : 0;
  const fov = (360 / Math.PI) * Math.atan(0.75 * 2 ** (1 - z));
  return Math.min(120, Math.max(30, fov));
}

export function viewPlan(lane, heading = 0, pitch = 0, zoom = 0) {
  const baseHeading = wrapHeading(heading);
  const posePitch = Number(pitch) || 0;
  const poseZoom = Number(zoom) || 0;
  const ringPitch = lane === "object" ? 0 : posePitch;
  const ringFov = lane === "object" ? FACE_FOV : thumbnailFov(poseZoom);
  const views = [];
  for (let offset = 0; offset < 4; offset += 1) {
    views.push({ yaw: wrapHeading(baseHeading + offset * 90), pitch: ringPitch, fov: ringFov });
  }
  views.push({ yaw: baseHeading, pitch: 90, fov: FACE_FOV });
  views.push({ yaw: baseHeading, pitch: -90, fov: FACE_FOV });
  if (views.length !== FACE_COUNT) throw new Error("view_count");
  return views;
}

function queryNumber(value) {
  const number = Number(value);
  if (!Number.isFinite(number)) return "0";
  if (Number.isInteger(number)) return String(number);
  return String(Number(number.toPrecision(6)));
}

export function thumbnailUrl(panoId, yaw, pitch, fov, width = THUMB_SIZE, height = THUMB_SIZE) {
  const     params = new URLSearchParams({
    cb_client: "apiv3",
    output: "thumbnail",
    panoid: panoId,
    w: String(width),
    h: String(height),
    yaw: queryNumber(yaw),
    pitch: queryNumber(-Number(pitch)),
    thumbfov: String(Math.round(Math.min(120, Math.max(30, Number(fov) || 0)))),
  });
  return `${THUMBNAIL_ENDPOINT}?${params}`;
}

export function downsampleBox(rgb, width, height, size = FACE_SIZE) {
  if (!rgb || width <= 0 || height <= 0 || rgb.length < width * height * 3) {
    throw new Error("invalid_thumbnail");
  }
  if (width === size && height === size) return rgb.slice(0, size * size * 3);
  const out = new Uint8Array(size * size * 3);
  for (let y = 0; y < size; y += 1) {
    const y0 = Math.floor((y * height) / size);
    let y1 = Math.floor(((y + 1) * height) / size);
    if (y1 <= y0) y1 = y0 + 1;
    for (let x = 0; x < size; x += 1) {
      const x0 = Math.floor((x * width) / size);
      let x1 = Math.floor(((x + 1) * width) / size);
      if (x1 <= x0) x1 = x0 + 1;
      const count = (y1 - y0) * (x1 - x0);
      let sumR = 0, sumG = 0, sumB = 0;
      for (let py = y0; py < y1; py += 1) {
        const row = py * width * 3;
        for (let px = x0; px < x1; px += 1) {
          const index = row + px * 3;
          sumR += rgb[index];
          sumG += rgb[index + 1];
          sumB += rgb[index + 2];
        }
      }
      const dest = (y * size + x) * 3;
      out[dest] = Math.floor(sumR / count);
      out[dest + 1] = Math.floor(sumG / count);
      out[dest + 2] = Math.floor(sumB / count);
    }
  }
  return out;
}

export function decodeJpegRgb(payload) {
  const decoded = jpegJs.decode(payload, { useTArray: true, formatAsRGBA: true });
  const pixels = decoded.data;
  const rgb = new Uint8Array(decoded.width * decoded.height * 3);
  for (let i = 0, o = 0; i < pixels.length; i += 4, o += 3) {
    rgb[o] = pixels[i];
    rgb[o + 1] = pixels[i + 1];
    rgb[o + 2] = pixels[i + 2];
  }
  return { rgb, width: decoded.width, height: decoded.height };
}

async function fetchThumbnail(url) {
  let lastError = null;
  for (let attempt = 0; attempt < 3; attempt += 1) {
    try {
      const response = await fetch(url, {
        headers: { "User-Agent": VIEW_USER_AGENT },
        cf: { cacheTtl: 0, cacheEverything: false },
      });
      if (!response.ok) throw new Error("view_unavailable");
      const payload = new Uint8Array(await response.arrayBuffer());
      if (!payload.length) throw new Error("view_unavailable");
      return payload;
    } catch (error) {
      lastError = error;
      if (attempt < 2) await new Promise((resolve) => setTimeout(resolve, 400 * (attempt + 1)));
    }
  }
  throw lastError || new Error("view_unavailable");
}

export async function renderStreetFaces(location, fetchBytes = fetchThumbnail) {
  const panoId = location.panoId || location.assetId || location.asset_id;
  if (!usesStreetViews(panoId)) throw new Error("not_a_street_pano");
  const lane = location.lane || "scene";
  const faces = new Uint8Array(FACES_BYTES);
  const views = viewPlan(lane, location.heading || 0, location.pitch || 0, location.zoom || 0);
  let offset = 0;
  for (const view of views) {
    const jpeg = await fetchBytes(thumbnailUrl(panoId, view.yaw, view.pitch, view.fov));
    const { rgb, width, height } = decodeJpegRgb(jpeg);
    const face = downsampleBox(rgb, width, height);
    faces.set(face, offset);
    offset += face.length;
  }
  if (offset !== FACES_BYTES) throw new Error("invalid_thumbnail");
  return faces;
}

export async function renderLocationFaces(location, seedFaces) {
  const panoId = location.panoId || location.assetId || location.asset_id;
  if (usesStreetViews(panoId)) return renderStreetFaces(location);
  return seedFaces();
}

export function streetLeaseCap(lane, pace) {
  return leaseCap(lane, pace, "browser");
}

export function leaseCap(lane, pace, client = "browser") {
  const table = client === "cli" ? CLI_LEASE_CAP : STREET_LEASE_CAP;
  const caps = table[lane] || table.scene;
  return caps[pace] || caps.medium;
}

File: community/pano.py
"""Street View view geometry matching VISION.app, without copying its models.

Scene processing uses the four-view compass (heading + 0/90/180/270) at the
saved pose pitch and thumbnail FOV from zoom. Object processing uses the
six-face cube (those four headings at pitch 0, plus zenith and nadir) at 90°
FOV. Community still downsamples each view to a 16×16 integer face so the
existing community-visual-v1 descriptor can be recomputed. Imagery is fetched
ephemerally and never written to disk, D1, or R2.

Invented catalog IDs (Prototype, CommunityPano, synthetic, wikimedia) keep the
identity-seed extractor so local tests do not hit the network.
"""

from __future__ import annotations

import io
import math
import time
import urllib.error
import urllib.parse
import urllib.request

from .features import FACE_COUNT, FACE_SIZE, FACES_BYTES, render_faces


THUMB_SIZE = 64
FACE_FOV = 90.0
THUMBNAIL_ENDPOINT = "https://geo0.ggpht.com/cbk"
VIEW_USER_AGENT = "VISION-Community-view/1"
INVENTED_PREFIXES = ("synthetic:", "Prototype", "CommunityPano", "wikimedia:")
QUERY_VIEW_CAP = 4
# Browser processing proxies thumbnails through the Worker (CORS), so leases stay small.
STREET_LEASE_CAP = {
    "scene": {"slow": 1, "medium": 4, "max": 8},
    "object": {"slow": 1, "medium": 2, "max": 4},
}
# CLI fetches Street View on the volunteer machine; larger exclusive batches are safe.
CLI_LEASE_CAP = {
    "scene": {"slow": 16, "medium": 64, "max": 128},
    "object": {"slow": 8, "medium": 32, "max": 64},
}


def lease_cap(lane: str, pace: str, client: str = "browser") -> int:
    table = CLI_LEASE_CAP if client == "cli" else STREET_LEASE_CAP
    caps = table.get(lane) or table["scene"]
    return int(caps.get(pace) or caps["medium"])


class ViewError(Exception):
    def __init__(self, code: str = "view_unavailable"):
        super().__init__(code)
        self.code = code


def uses_street_views(pano_id: str | None) -> bool:
    if not isinstance(pano_id, str) or not pano_id.strip():
        return False
    identity = pano_id.strip()
    if "maps.googleapis.com" in identity or identity.startswith("http"):
        return False
    return not identity.startswith(INVENTED_PREFIXES)


def wrap_heading(heading: float) -> float:
    return float(heading) % 360.0


def thumbnail_fov(zoom: float) -> float:
    try:
        zoom_value = float(zoom)
    except (TypeError, ValueError):
        zoom_value = 0.0
    if not math.isfinite(zoom_value):
        zoom_value = 0.0
    fov = (360.0 / math.pi) * math.atan(0.75 * (2.0 ** (1.0 - zoom_value)))
    return min(120.0, max(30.0, fov))


def view_plan(lane: str, heading: float = 0, pitch: float = 0, zoom: float = 0) -> list[dict]:
    """Return the six faces stored in a community-visual-v1 buffer.

    Scene compass views follow VISION four-view (pose pitch + zoom FOV).
    Object compass views follow the VISION object cube (pitch 0, FOV 90).
    Faces 4 and 5 are zenith and nadir at FOV 90 for both lanes so the
    six-face descriptor stays fully populated from real pixels.
    """
    base_heading = wrap_heading(heading or 0)
    pose_pitch = float(pitch or 0)
    pose_zoom = float(zoom or 0)
    if lane == "object":
        ring_pitch = 0.0
        ring_fov = FACE_FOV
    else:
        ring_pitch = pose_pitch
        ring_fov = thumbnail_fov(pose_zoom)
    views = []
    for offset in range(4):
        views.append(
            {
                "yaw": wrap_heading(base_heading + offset * 90.0),
                "pitch": ring_pitch,
                "fov": ring_fov,
            }
        )
    views.append({"yaw": base_heading, "pitch": 90.0, "fov": FACE_FOV})
    views.append({"yaw": base_heading, "pitch": -90.0, "fov": FACE_FOV})
    if len(views) != FACE_COUNT:
        raise RuntimeError("view_count")
    return views


def thumbnail_url(pano_id: str, yaw: float, pitch: float, fov: float, *, width: int = THUMB_SIZE, height: int = THUMB_SIZE) -> str:
    query = urllib.parse.urlencode(
        {
            "cb_client": "apiv3",
            "output": "thumbnail",
            "panoid": pano_id,
            "w": str(width),
            "h": str(height),
            "yaw": _query_number(yaw),
            "pitch": _query_number(-float(pitch)),
            "thumbfov": str(int(round(min(120.0, max(30.0, float(fov)))))),
        }
    )
    return f"{THUMBNAIL_ENDPOINT}?{query}"


def _query_number(value: float) -> str:
    number = float(value)
    if number.is_integer():
        return str(int(number))
    return f"{number:.6g}"


def downsample_box(rgb: bytes, width: int, height: int, size: int = FACE_SIZE) -> bytes:
    if width <= 0 or height <= 0 or len(rgb) < width * height * 3:
        raise ViewError("invalid_thumbnail")
    if width == size and height == size:
        return bytes(rgb[: size * size * 3])
    out = bytearray(size * size * 3)
    for y in range(size):
        y0 = y * height // size
        y1 = (y + 1) * height // size
        if y1 <= y0:
            y1 = y0 + 1
        for x in range(size):
            x0 = x * width // size
            x1 = (x + 1) * width // size
            if x1 <= x0:
                x1 = x0 + 1
            count = (y1 - y0) * (x1 - x0)
            sum_r = sum_g = sum_b = 0
            for py in range(y0, y1):
                row = py * width * 3
                for px in range(x0, x1):
                    index = row + px * 3
                    sum_r += rgb[index]
                    sum_g += rgb[index + 1]
                    sum_b += rgb[index + 2]
            dest = (y * size + x) * 3
            out[dest] = sum_r // count
            out[dest + 1] = sum_g // count
            out[dest + 2] = sum_b // count
    return bytes(out)


def decode_jpeg_rgb(payload: bytes) -> tuple[bytes, int, int]:
    try:
        from PIL import Image
    except ImportError as error:
        raise ViewError("jpeg_decoder_missing") from error
    try:
        image = Image.open(io.BytesIO(payload)).convert("RGB")
    except Exception as error:
        raise ViewError("invalid_thumbnail") from error
    return image.tobytes(), image.width, image.height


def fetch_thumbnail(url: str, *, attempts: int = 3) -> bytes:
    last_error: Exception | None = None
    for attempt in range(max(1, attempts)):
        request = urllib.request.Request(url, headers={"User-Agent": VIEW_USER_AGENT})
        try:
            with urllib.request.urlopen(request, timeout=20) as response:
                payload = response.read()
                if getattr(response, "status", 200) != 200 or not payload:
                    raise ViewError("view_unavailable")
                return payload
        except ViewError as error:
            last_error = error
        except (urllib.error.URLError, TimeoutError, OSError) as error:
            last_error = error
        if attempt + 1 < attempts:
            time.sleep(0.4 * (attempt + 1))
    raise ViewError("view_unavailable") from last_error


def render_street_faces(location: dict, *, fetch=fetch_thumbnail) -> bytes:
    pano_id = location.get("panoId") or location.get("assetId") or location.get("asset_id")
    if not uses_street_views(pano_id):
        raise ViewError("not_a_street_pano")
    lane = location.get("lane") or "scene"
    faces = bytearray()
    for view in view_plan(lane, location.get("heading") or 0, location.get("pitch") or 0, location.get("zoom") or 0):
        payload = fetch(thumbnail_url(pano_id, view["yaw"], view["pitch"], view["fov"]))
        rgb, width, height = decode_jpeg_rgb(payload)
        faces.extend(downsample_box(rgb, width, height))
    if len(faces) != FACES_BYTES:
        raise ViewError("invalid_thumbnail")
    return bytes(faces)


def render_location_faces(location: dict, *, fetch=fetch_thumbnail) -> bytes:
    pano_id = location.get("panoId") or location.get("assetId") or location.get("asset_id")
    capture = location.get("capture") or "unknown"
    lane = location.get("lane") or "scene"
    model = location.get("model") or "community-visual-v1"
    if uses_street_views(pano_id):
        return render_street_faces(location, fetch=fetch)
    return render_faces(pano_id, capture, lane, model)


def location_from_row(row) -> dict:
    if isinstance(row, dict):
        mapping = row
    else:
        mapping = {key: row[key] for key in row.keys()}
    return {
        "panoId": mapping.get("asset_id") or mapping.get("panoId") or mapping.get("assetId"),
        "assetId": mapping.get("asset_id") or mapping.get("assetId") or mapping.get("panoId"),
        "capture": mapping.get("capture") or "unknown",
        "lane": mapping.get("lane") or "scene",
        "model": mapping.get("model") or "community-visual-v1",
        "heading": mapping.get("heading") or 0,
        "pitch": mapping.get("pitch") or 0,
        "zoom": mapping.get("zoom") or 0,
    }
